# Проект по SQL

# Описание проекта:

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Цель исследования:
Сформулировать ценностное предложение для нового продукта.

# Описание данных: 

Таблица **books** содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица **authors** содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

Таблица **publishers** содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;
- Таблица ratings
- Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица **reviews** содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

# Схема данных

In [1]:
from IPython.display import Image
Image(url = 'https://pictures.s3.yandex.net/resources/scheme_1589269096.png')

# План работы:


- Подготовить данные для составления SQL-запросов
- Ознакомиться с данными и схеммой данных
- Выполнить задания 
- Подвести итоги

# Задания:

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

# Подготовка данных:

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
# Функция выполнения sql запроса

def func(sql):
    return pd.read_sql(sql,con=engine)

Далее подготовим все, необходимые нам в работе, таблицы:

In [4]:
sql_books = '''
SELECT *
FROM books
LIMIT 10
'''
books = func(sql_books)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [5]:
sql_ratings = '''
SELECT *
FROM ratings
LIMIT 10
'''
ratings = func(sql_ratings)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [6]:
sql_authors = '''
SELECT *
FROM authors
LIMIT 10
'''
authors = func(sql_authors)
authors

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [7]:
sql_publishers = '''
SELECT *
FROM publishers
LIMIT 10
'''
publishers = func(sql_publishers)
publishers

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [8]:
sql_reviews = '''
SELECT *
FROM reviews
LIMIT 10
'''
reviews = func(sql_reviews)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Cколько книг вышло после 1 января 2000 года;

In [9]:
first_exercise = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'

'''
func(first_exercise)


count
0    819

**Выводы:**
После 1 января 2000 года вышло 819 новых книг

## Количество обзоров и средняя оценка для каждой книги

- Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [14]:
second_exercise = '''
SELECT  b.book_id,
        b.title,
        COUNT(DISTINCT(rev.review_id)),
        AVG(rating)
FROM books AS b
LEFT OUTER JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT OUTER JOIN ratings AS r ON b.book_id = r.book_id
GROUP BY b.book_id

'''
func(second_exercise)


book_id                                              title  count  \
0          1                                       'Salem's Lot      2   
1          2                 1 000 Places to See Before You Die      1   
2          3  13 Little Blue Envelopes (Little Blue Envelope...      3   
3          4  1491: New Revelations of the Americas Before C...      2   
4          5                                               1776      4   
..       ...                                                ...    ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)      3   
996      997                        Xenocide (Ender's Saga  #3)      3   
997      998                                    Year of Wonders      4   
998      999                        You Suck (A Love Story  #2)      2   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...      4   

          avg  
0    3.666667  
1    2.500000  
2    4.666667  
3    4.500000  
4    4.000000  
..        ...  
995  3.666667  
996  3.400000  
997  3.200000  
998  4.500000  
999  3.833333  

[1000 rows x 4 columns]

**Выводы:**

Сформировали таблицу, в которой содержатся:
- Количество обзоров
- Средний рейтинг
Для каждой книги.
Анализ информации из данной таблицы может содержать множество идей для формирования ценностного предложения.
Для нашего сервиса было бы полезно изучить его более подробно.

## Издательство, которое выпустило наибольшее число книг толще 50 страниц 
— так вы исключите из анализа брошюры;

In [11]:
third_exercise = '''
SELECT 
        p.publisher,
        COUNT((p.publisher_id)) AS cnt
FROM books AS b
LEFT OUTER JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY b.publisher_id, p.publisher
ORDER BY cnt DESC
LIMIT 1
'''
func(third_exercise)


publisher  cnt
0  Penguin Books   42

**Вывод:**
 - Издательство, которое выпустило наибольшее число книг, которые толще 50 страниц - "Penguin Books"

## Автор с самой высокой средней оценкой книг
— учитывайте только книги с 50 и более оценками;

In [12]:
fourth_exercise = '''

WITH books_avg_rat AS (SELECT books.book_id,
                              AVG(ratings.rating) AS rat
                       FROM books
                       LEFT OUTER JOIN ratings ON books.book_id = ratings.book_id
                       GROUP BY books.book_id
                       HAVING COUNT(ratings.book_id) >= 50)

SELECT  authors.author,
        ROUND(AVG(books_avg_rat.rat),2)
FROM books_avg_rat 
LEFT OUTER JOIN books ON books_avg_rat.book_id = books.book_id
LEFT OUTER JOIN authors ON books.author_id = authors.author_id
GROUP BY authors.author
LIMIT 1
'''
func(fourth_exercise)

author  round
0  J.K. Rowling/Mary GrandPré   4.28

**Вывод:**

- Автор с самой высокой средней оценкой книг: 'J.K. Rowling/Mary GrandPré	'\
- Средняя оценка книг этого автора составила: '4.28'

## Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [13]:
fifth_exercise = '''

WITH pop_users AS (SELECT username,
                          COUNT(ratings.rating)
                              
                   FROM ratings
                   GROUP BY username
                   HAVING COUNT(ratings.rating) > 48)
                   
SELECT AVG(t.cnt)
FROM (SELECT  pop_users.username, 
      COUNT(review_id) AS cnt
FROM reviews
INNER JOIN pop_users ON reviews.username = pop_users.username
GROUP BY pop_users.username) AS t
'''
func(fifth_exercise)

avg
0  24.0

**Вывод:**
В среднем, на одного заинтересованного пользователя (который поставил более 48 оценок), приходится по 24 обзора. 

## Итоги

- После 1 января 2000 года вышло 819 новых книг.

- Сформировали таблицу, в которой содержатся:

Количество обзоров и средний рейтинг Для каждой книги.\
Анализ информации из данной таблицы может содержать множество идей для формирования ценностного предложения. Для нашего сервиса было бы полезно изучить его более подробно.

- Издательство, которое выпустило наибольшее число книг, которые толще 50 страниц - "Penguin Books"

- Автор с самой высокой средней оценкой книг: 'J.K. Rowling/Mary GrandPré	'
- Средняя оценка книг этого автора составила: '4.28'
-  В среднем, на одного заинтересованного пользователя (который поставил более 48 оценок), приходится по 24 обзора.